crates/sui-network/src/discovery/mod.rs

```diff
@@ -0,0 +1,448 @@
+            }
+
+            Err(RecvError::Closed) => {
+                panic!("PeerEvent channel shouldn't be able to be closed");
+            }
+
+            Err(RecvError::Lagged(_)) => {
+                trace!("State-Sync fell behind processing PeerEvents");
+            }
+        }
+    }
+
+    fn handle_tick(&mut self, _now: std::time::Instant, now_unix: u64) {
+        self.update_our_info_timestamp(now_unix);
+
+        self.tasks
+            .spawn(query_connected_peers_for_their_known_peers(
+                self.network.clone(),
+                self.state.clone(),
+            ));
+
+        // Cull old peers older than a day
+        self.state
+            .write()
+            .unwrap()
+            .known_peers
+            .retain(|_k, v| now_unix.saturating_sub(v.timestamp_ms) < ONE_DAY_MILLISECONDS);
+
+        // Clean out the pending_dials
+        self.pending_dials.retain(|_k, v| !v.is_finished());
+
+        // Spawn some dials
+        let state = self.state.read().unwrap();
+        let mut eligible = state
+            .known_peers
+            .clone()
+            .into_iter()
+            .filter(|(peer_id, info)| {
+                peer_id != &self.network.peer_id() &&
+                (!info.addresses.is_empty() || info.external_socket_address.is_some()) // Peer has addresses we can dial
+                && !state.connected_peers.contains_key(peer_id) // We're not already connected
+                && !self.pending_dials.contains_key(peer_id) // There is no pending dial to this node
+            })
+            .collect::<Vec<_>>();
+
+        // randomize the order
+        let mut rng = <rand::rngs::StdRng as rand::SeedableRng>::from_entropy();
+        rand::seq::SliceRandom::shuffle(eligible.as_mut_slice(), &mut rng);
+
+        // No need to connect to any more peers if we're already connected to a bunch
+        let number_of_connections = self.state.read().unwrap().connected_peers.len();
+        let number_to_dial = std::cmp::min(
+            eligible.len(),
+            MAX_NUMBER_OF_CONNECTIONS.saturating_sub(number_of_connections),
+        );
+
+        for (peer_id, info) in eligible.into_iter().take(number_to_dial) {
+            let abort_handle = self
+                .tasks
+                .spawn(try_to_connect_to_peer(self.network.clone(), info));
+            self.pending_dials.insert(peer_id, abort_handle);
+        }
+
+        // If we aren't connected to anything and we aren't presently trying to connect to anyone
+        // we need to try the seed peers
+        if state.connected_peers.is_empty()
+            && self.pending_dials.is_empty()
+            && !self.config.seed_peers.is_empty()
+        {
+            let abort_handle = self.tasks.spawn(try_to_connect_to_seed_peers(
+                self.network.clone(),
+                self.config.seed_peers.clone(),
+            ));
+
+            // Use the all zero's peer_id as a marker for dialing seed peers
```

Comment:
this makes me feel mildly icky. do you think it adds too much code bloat to make the key type an enum that can be either `Peer(PeerId)` or `SeedPeers`?

Reply:
Yeah that does make me feel a bit iky too...i've got another idea to fix this though, which i'll do instead.